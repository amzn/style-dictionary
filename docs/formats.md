# Formats

Formats are one of the ways to create files that act as interfaces for your style dictionary. For example, you want to be able to
use your style dictionary in CSS. You can use the `css/variables` template which will create a CSS file with variables from
your style dictionary. You can define custom formats with the [`registerFormat`](api.md#registerformat).

Templates and Formats serve the same purpose: use your style dictionary as data to build a file. You use formats in your config
file under platforms > [platform] > files > [file]

```json
{
  "source": ["properties/**/*.json"],
  "platforms": {
    "css": {
      "transformGroup": "css",
      "files": [
        {
          "template": "css/variables",
          "destination": "variables.css"
        }
      ]
    }
  }
}
```


>*__How are Templates different than Formats?__*

>Mainly syntactic sugar; anything you can do in a Template you can do in a Format. Use whichever is easier for you to write. We find
that Templates are good if you have a lot of boilerplate code around where the style dictionary will go (like writing ObjectiveC files).
Formats are better if there is little to no boilerplate code like a flat SCSS variables file.

----

## Pre-defined Formats

[lib/common/formats.js](https://github.com/amzn/style-dictionary/blob/master/lib/common/formats.js)

### css/variables


Creates a CSS file with variable definitions based on the style dictionary

**Example**
```css
:root {
  --color-background-base: #f0f0f0;
  --color-background-alt: #eeeeee;
}
```

* * *

### scss/variables


Creates a SCSS file with variable definitions based on the style dictionary

**Example**
```scss
$color-background-base: #f0f0f0;
$color-background-alt: #eeeeee;
```

* * *

### scss/icons


Creates a SCSS file with variable definitions and helper classes for icons

**Example**
```scss
$content-icon-email: '\E001';
.icon.email:before { content:$content-icon-email; }
```

* * *

### less/variables


Creates a LESS file with variable definitions based on the style dictionary

**Example**
```less
@color-background-base: #f0f0f0;
@color-background-alt: #eeeeee;
```

* * *

### less/icons


Creates a LESS file with variable definitions and helper classes for icons

**Example**
```less
@content-icon-email: '\E001';
.icon.email:before { content:@content-icon-email; }
```

* * *

### javascript/module


Creates a CommonJS module with the whole style dictionary

**Example**
```js
module.exports = {
  color: {
    base: {
       red: {
         value: '#ff000'
       }
    }
  }
}
```

* * *

### javascript/object


Creates a JS file a global var that is a plain javascript object of the style dictionary.
Name the variable by adding a 'name' attribute on the file object in your config.

**Example**
```js
var StyleDictionary = {
  color: {
    base: {
       red: {
         value: '#ff000'
       }
    }
  }
}
```

* * *

### javascript/umd


Creates a [UMD](https://github.com/umdjs/umd) module of the style
dictionary. Name the module by adding a 'name' attribute on the file object
in your config.

**Example**
```js
(function(root, factory) {
  if (typeof module === "object" && module.exports) {
    module.exports = factory();
  } else if (typeof exports === "object") {
    exports["_styleDictionary"] = factory();
  } else if (typeof define === "function" && define.amd) {
    define([], factory);
  } else {
    root["_styleDictionary"] = factory();
  }
}(this, function() {
  return {
    "color": {
      "red": {
        "value": "#FF0000"
      }
    }
  };
}))
```

* * *

### javascript/es6


Creates a ES6 module of the style dictionary. You can filter the style dictionary
to only export properties of a certain type by adding a 'filter' attribute on the
file object in the config.

```json
{
  "platforms": {
    "js": {
      "transformGroup": "js",
      "files": [
        {
          "format": "javascript/es6",
          "destination": "colors.js",
          "filter": {
            "attributes": {
              "category": "color"
            }
          }
        }
      ]
    }
  }
}
```

**Example**
```js
export const ColorBackgroundBase = '#ffffff';
export const ColorBackgroundAlt = '#fcfcfcfc';
```

* * *

### json


Creates a JSON file of the style dictionary.

**Example**
```json
{
  "color": {
    "base": {
       "red": {
         "value": "#ff000"
       }
    }
  }
}
```

* * *

### json/asset


Creates a JSON file of just the assets defined in the style dictionary.

**Example**
```js
{
  "asset": {
    "image": {
       "logo": {
         "value": "assets/logo.png"
       }
    }
  }
}
```

* * *

### sketch/palette


Creates a sketchpalette file of all the base colors

**Example**
```json
{
  "compatibleVersion": "1.0",
  "pluginVersion": "1.1",
  "colors": [
    "#ffffff",
    "#ff0000",
    "#fcfcfc"
  ]
}
```

* * *
