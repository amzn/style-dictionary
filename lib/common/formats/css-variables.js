function formatter(dictionary, config) {
  const header = config.header || '';
  const body = dictionary.allProperties.map(mapProp.bind(this, config));
  return `${header}:root {\n${body.join("\n")}\n}\n`;
}

function mapProp(config, prop) {
  const comment = (prop.comment) ? ` /* ${prop.comment} */` : '';
  const varName = makePropCSSVar(prop.name);
  const value = setValue(prop, config);
  let token = ` ${varName}: ${value};${comment}`;

  if (config.dark && prop.attributes && prop.attributes.dark) {
    const darkProp = generateDarkToken(prop);
    const darkValue = setValue(darkProp, config);
    token += `;\n ${varName}-dark: ${darkValue};`;
  }

  return token;
}

function setValue(prop, config) {
  if (typeof prop.value === "object") {
    throwObjectError(prop);
  }

  const nVal = prop.value;
  const oVal = prop.original && prop.original.value;
  if (config.deep && nVal !== oVal && isAlias(oVal)) {
    return useReferenceValue(oVal, config.prefix);
  }
  return typeIsString(prop) ? `"${nVal}"` : nVal;
}

function generateDarkToken(prop) {
  let returnProp = prop;
  if (prop.attributes && prop.attributes.dark) {
    returnProp = {
      name: prop.name+'-dark',
      value: prop.attributes.dark.value,
      original: { value: prop.attributes.dark.original },
    };
  }
  return returnProp;
}

function throwObjectError(prop) {
  let message = `"${prop.name}" has an original value of "${prop.original.value}". \n`;
  message += "This points to an object. ";
  message += "Reference the object's \"value\" key if it's an alias \n";
  throw new Error(message);
}

function isAlias(str) {
  // is string and matches "{text.value}"
  return !!(typeof str === "string" && str.match(/(^{.*\.value}$)/gm));
}

function typeIsString(prop) {
  return !!(prop.attributes && prop.attributes["data-type"] === "string");
}

function makePropCSSVar(name) {
  return `--${name}`;
}

function useReferenceValue(value, prefix) {
  const head = prefix ? `var(--${prefix}-` : "var(--";
  return `${head}${value.replace(/\./g, "-").replace(/({|-value})/g, "")})`;
}

module.exports = formatter;
