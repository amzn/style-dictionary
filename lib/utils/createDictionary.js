/*
 * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

const flattenProperties = require('./flattenProperties');
const getReferences = require('./references/getReferences');
const usesReference = require('./references/usesReference');

/**
 *
 * @typedef Dictionary
 * @property {Object} properties
 * @property {Array} allProperties
 * @property {Dictionary.getReferences} getReferences
 * @property {Dictionary.usesReference} usesReference
 */

/**
 * Creates the dictionary object that is passed to formats and actions.
 * @param {Object} args
 * @param {Object} args.properties
 * @returns {Dictionary}
 */
function createDictionary({ properties }) {
  return {
    properties: properties,
    allProperties: flattenProperties( properties ),
    getReferences: getReferences,
    usesReference: usesReference
  }
}

module.exports = createDictionary;