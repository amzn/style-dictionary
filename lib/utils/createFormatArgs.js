/*
 * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

const deepExtend = require('./deepExtend');

function createFormatArgs({ dictionary, platform, file = {} }) {
  const {allProperties, properties, allTokens, tokens, usesReference, getReferences} = dictionary;
  // This will merge platform and file-level configuration
  // where the file configuration takes precedence
  const {options} = platform;
  file = deepExtend([{}, {options}, file]);

  return {
    dictionary,
    usesReference,
    getReferences,
    allProperties,
    properties,
    // adding tokens and allTokens as the new way starting in v3,
    // keeping properties and allProperties around for backwards-compatibility
    allTokens,
    tokens,
    platform,
    file,
    options: file.options || {}
  }
}

module.exports = createFormatArgs;