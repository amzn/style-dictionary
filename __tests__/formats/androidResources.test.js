/*
 * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */
import { expect } from 'chai';
import formats from '../../lib/common/formats.js';
import createDictionary from '../../lib/utils/createDictionary.js';
import createFormatArgs from '../../lib/utils/createFormatArgs.js';

const tokens = {
  size: {
    font: {
      small: {
        value: '12rem',
        original: {
          value: '12px',
        },
        name: 'size-font-small',
        attributes: {
          category: 'size',
          type: 'font',
          item: 'small',
        },
        path: ['size', 'font', 'small'],
      },
      large: {
        value: '18rem',
        original: {
          value: '18px',
        },
        name: 'size-font-large',
        attributes: {
          category: 'size',
          type: 'font',
          item: 'large',
        },
        path: ['size', 'font', 'large'],
      },
    },
  },
  color: {
    base: {
      red: {
        value: '#ff0000',
        comment: 'comment',
        original: {
          value: '#FF0000',
          comment: 'comment',
        },
        name: 'color-base-red',
        attributes: {
          category: 'color',
          type: 'base',
          item: 'red',
        },
        path: ['color', 'base', 'red'],
      },
    },
    white: {
      value: '#ffffff',
      original: {
        value: '#ffffff',
      },
      name: 'color-white',
      attributes: {
        category: 'color',
        type: 'white',
      },
      path: ['color', 'white'],
    },
  },
};

const customTokens = {
  backgroundColor: {
    secondary: {
      name: 'backgroundColorSecondary',
      value: '#F2F3F4',
      attributes: {
        category: 'backgroundColor',
      },
    },
  },
  fontColor: {
    primary: {
      name: 'fontColorPrimary',
      value: '#000000',
      attributes: {
        category: 'fontColor',
      },
    },
  },
};

const format = formats['android/resources'];
const file = {
  destination: '__output/',
  format: 'android/resources',
};

const dictionary = createDictionary({ tokens });
const customDictionary = createDictionary({ tokens: customTokens });

describe('formats', () => {
  describe(`android/resources`, () => {
    it('should match default snapshot', async () => {
      await expect(
        format(
          createFormatArgs({
            dictionary,
            file,
            platform: {},
          }),
          {},
          file,
        ),
      ).to.matchSnapshot();
    });

    it('with resourceType override should match snapshot', async () => {
      const file = { resourceType: 'dimen' };
      await expect(
        format(
          createFormatArgs({
            dictionary,
            file,
            platform: {},
          }),
          {},
          file,
        ),
      ).to.matchSnapshot();
    });

    it('with resourceMap override should match snapshot', async () => {
      const file = {
        resourceMap: {
          color: 'color',
          fontColor: 'color',
          backgroundColor: 'color',
        },
      };
      await expect(
        format(
          createFormatArgs({
            dictionary: customDictionary,
            file,
            platform: {},
          }),
          {},
          file,
        ),
      ).to.matchSnapshot();
    });
  });
});
