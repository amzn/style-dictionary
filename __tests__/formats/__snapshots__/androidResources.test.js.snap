// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formats android/resources should match default snapshot 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>

<!--
  Do not edit directly
  Generated on Sat, 01 Jan 2000 00:00:00 GMT
-->
<resources>
  <dimen name=\\"size-font-small\\">12rem</dimen>
  <dimen name=\\"size-font-large\\">18rem</dimen>
  <color name=\\"color-base-red\\">#ff0000</color><!-- comment -->
  <color name=\\"color-white\\">#ffffff</color>
  
</resources>
"
`;

exports[`formats android/resources with resourceMap override should match snapshot 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>

<!--
  Do not edit directly
  Generated on Sat, 01 Jan 2000 00:00:00 GMT
-->
<resources>
  <color name=\\"backgroundColorSecondary\\">#F2F3F4</color>
  <color name=\\"fontColorPrimary\\">#000000</color>
  
</resources>
"
`;

exports[`formats android/resources with resourceType override should match snapshot 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>

<!--
  Do not edit directly
  Generated on Sat, 01 Jan 2000 00:00:00 GMT
-->
<resources>
  <dimen name=\\"size-font-small\\">12rem</dimen>
  <dimen name=\\"size-font-large\\">18rem</dimen>
  <dimen name=\\"color-base-red\\">#ff0000</dimen><!-- comment -->
  <dimen name=\\"color-white\\">#ffffff</dimen>
  
</resources>
"
`;
